ctru-rs/src/test_runner.rs

```diff
@@ -0,0 +1,135 @@
+//! Custom test runner for building/running unit tests on the 3DS.
+
+extern crate test;
+
+use std::io;
+
+use test::{ColorConfig, Options, OutputFormat, RunIgnored, TestDescAndFn, TestFn, TestOpts};
+
+use crate::console::Console;
+use crate::gfx::Gfx;
+use crate::services::hid::{Hid, KeyPad};
+
+/// A custom runner to be used with `#[test_runner]`. This simple implementation
+/// runs all tests in series, "failing" on the first one to panic (really, the
+/// panic is just treated the same as any normal application panic).
+pub(crate) fn test_runner(tests: &[&TestDescAndFn]) {
+    crate::init();
+
+    let gfx = Gfx::default();
+    let hid = Hid::init().unwrap();
+
+    let mut top_screen = gfx.top_screen.borrow_mut();
+    top_screen.set_wide_mode(true);
+    let _console = Console::init(top_screen);
+
+    // Start printing from the top left
+    print!("\x1b[1;1H");
+
+    // TODO: it would be nice to have a way of specifying argv to make these
+    // configurable at runtime, but I can't figure out how to do it easily,
+    //  so for now, just hardcode everything.
+    let opts = TestOpts {
+        list: false,
+        filters: Vec::new(),
+        filter_exact: false,
+        // Forking is not supported
+        force_run_in_process: true,
+        exclude_should_panic: false,
+        run_ignored: RunIgnored::No,
+        run_tests: true,
+        // Benchmarks are not supported
+        bench_benchmarks: false,
+        logfile: None,
+        nocapture: false,
+        // TODO: color doesn't work because of TERM/TERMINFO.
+        // With RomFS we might be able to fake this out nicely...
+        color: ColorConfig::AlwaysColor,
```

Comment:
How would this work? Also, shouldn't we be disabling color here instead of force-enabling it?

Reply:
> Also, shouldn't we be disabling color here instead of force-enabling it?

Perhaps yes, or just setting it to "auto" which will I think leave it disabled.

> How would this work? 

The `test` implementation, on unix-like OSes, uses the `TERM` env var to learn what terminal type is being used, and `TERMINFO` to look up that terminal type in the [`terminfo` database](https://man7.org/linux/man-pages/man5/terminfo.5.html).

I think if we set the variables appropriately and add a terminfo file to the RomFS, we can get the test library to treat the console like a normal ANSI console, which would include hopefully include nice colored output. This is definitely a "future enhancement" kind of thing, though and shouldn't block this PR as far as I'm concerned.